ADR/adr-0008-environment-provisioning.md

```diff
@@ -0,0 +1,464 @@
+
+Represents a request for a [DeploymentTarget]. The `phase` section shows the lifecycle phase of the [DTC].
+
+**Immutable object**: no
+
+**Scope**: namespace
+
+```yaml
+apiVersion: appstudio.redhat.com/v1alpha1
+kind: DeploymentTargetClaim
+metadata:
+    name: prod-dtc
+spec:
+    deploymentTargetClassName: isolation-level-namespace
+status:
+    phase: Bound
+```
+
+Phases:
+
+**Pending:** DTC wait for the binding controller or user to bind it with a DT that satisfies it.
+
+**Bound:** The DTC was bounded to a DT that satisfies it by the binding controller.
+
+**Lost:** The DTC lost its bounded DT. The DT doesn't exist anymore because it got deleted.
+
+#### DeploymentTargetClass (DTCLS)
+
+Referred from a [DeploymentTarget] and [DeploymentTargetClaim]. Defines [DeploymentTarget] properties that
+should be abstracted from the controller/user that creates a [DTC] and wants a [DT] to be provisioned
+automatically for it.
+
+In the example below you can see a class that represents a [DT] that grants the requestor access to
+a namespace. The requestor isn’t aware of the actual location the [DT] is going to be provisioned. The
+parameters section can be used to forward additional information to the provisioner. The
+reclaimPolicy field will tell the provisioner what to do with the [DT] once its corresponding [DTC] is
+deleted, the values can be Retain or Delete.
+
+**Immutable object**: yes
+
+**Scope**: cluster
+
+```yaml
+apiVersion: appstudio.redhat.com/v1alpha1
+kind: DeploymentTargetClass
+metadata:
+    name: isolation-level-namespace
+spec:
+    provisioner: appstudio.redhat.com/devsandbox
+    parameters: {}
+    reclaimPolicy: Delete
+```
+
+#### Environment
+
+[Environment] objects refer to a [DTC] using the `deploymentTargetClaim`
+field. The environment controller will wait for the [DTC] to get to the
+`bound` phase, once it is bound, it will reach the [DT] and read the
+target's connection details from the kubernetesCredentials field and
+configure Argo/Gitops services to use them.
+
+**Immutable object**: no
+
+**Scope**: namespace
+
+```yaml
+apiVersion: appstudio.redhat.com/v1alpha1
+kind: Environment
+metadata:
+    name: prod
+spec:
+    displayName: “Production for Team A”
+    deploymentStrategy: AppStudioAutomated
+    parentEnvironment: staging
+    tags: [prod]
+    configuration:
+        target:
+            deploymentTargetClaim:
+                claimName: prod-dtc
+```
+
+### Controllers
+
+#### DeploymentTargetBinder
+
+Binds [DeploymentTargetClaim] to a [DeploymentTarget] that satisfies its requirements.
+
+It watches for [DTC] and objects and tries to find a matching [DT] for each one of them. In addition, it's responsible for updating the `Status` sections of the [DT] and [DTC] (it also watches [DTC] objects). It marks [DTC] objects that requires dynamic provisioning.
+
+A [DT] that was created dynamically for a specific [DTC] will always be attached to it.
+
+[DT] and [DTC] have one to one bindings.
```

Comment:
I think it make sense to have DTC only maps to one DT. Is there any case where we can have more than one DTC maps to the same DT?

Reply:
I don't have a concrete use case at the moment. We can perhaps imagine a use case where there is some single fixed cluster that we want to make available to multiple teams. We might hope to model the cluster as one DT and then let multiple DTCs be bound to it.

Or, instead maybe the single cluster should get multiple DTs created for it and the multiple DTCs should bind exclusively to each of those. This would at least encourage a model where the multiple DTCs do not share a Secret reference (of course, I guess they could all use the same Secret name - but at least the model encourages that kind of separation).

Reply:
At this point, I don't see a reason to have a one-to-many relationship between a DT and a DTC. If we want to use a different namespace on a target cluster, each namespace can be represented with its own DT.